src/common/datasource/src/util/io.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// modified from https://github.com/tokio-rs/tokio/blob/03912b9cf7141a5e27612a0b283d266637d2088f/tokio-util/src/io/sync_bridge.rs
+
+use std::io::Read;
+
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+#[derive(Debug)]
+pub struct SyncIoBridge<T> {
+    src: T,
+}
+
+impl<T: AsyncRead + Unpin> Read for SyncIoBridge<T> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let src = &mut self.src;
+        futures::executor::block_on(AsyncReadExt::read(src, buf))
```

Comment:
Not a good practice IMO. And I've checked two usages of this trait, it looks like the current implementation first get a sync/async `Read`, converts it to async via `convert_async_read` and then bridges it back to sync to use. Do you think we can avoid this?

Reply:
I looked at the `arrow-csv,` and it depends on the `csv` package, which only accepts reader implements `Read`. The OpenDAL only implements the `AsyncRead` trait, and we might need to create a PR for `arrow` to support the `AsyncRead` reader(maybe create a PR for `csv` first 🥲). I guess facing the same problem in the `arrow-json` package.

Reply:
>The OpenDAL only implements the AsyncRead trait

Does this [method](https://docs.rs/opendal/latest/opendal/struct.Operator.html#method.blocking) work?

Reply:
> > The OpenDAL only implements the AsyncRead trait
> 
> Does this [method](https://docs.rs/opendal/latest/opendal/struct.Operator.html#method.blocking) work?

It only works in the filesystem🥲

Reply:
Well... get it 😵‍💫. Will try to see how to support async io in csv/json reader